### PR TITLE
feat: implement template-based protoc plugin for interface generation

### PR DESCRIPTION
**Pull request by amery**
## Summary

- Implements template-based code generation system for protoc-gen-protomcp
- Adds example proto files demonstrating the generator output
- Establishes foundation for generating type-safe Go interfaces from protobuf definitions

## Details

This PR introduces the core protoc plugin that generates Go interfaces from Protocol Buffer service definitions. The implementation uses Go's text/template system for clean, maintainable code generation.

### Key Features

**Template-based Generation:**
- Modular template files for file structure, messages, fields, services, and RPCs
- Configurable interface naming patterns (e.g., `I%` prefix)
- Proper handling of proto3 optional fields without duplicate methods
- Organized imports with standard library first, followed by third-party packages

**Generated Code Capabilities:**
- Message interfaces with getter/setter methods that return errors for future validation
- Service interfaces with context-aware method signatures
- Full support for all proto3 field types including maps, repeated fields, and oneofs
- Clean separation of concerns with interface-first design

**Testing and Examples:**
- Comprehensive test utilities in `internal/testutils` for proto descriptor testing
- Working calculator service example with Add, Multiply, and Divide operations
- Build integration with `make examples` and `make examples-lint` targets
- buf configuration for proto linting with STANDARD rules

### Generated Code Example

```go
// Message interface with validation-ready setters
type IAddRequest interface {
    GetA() float64
    SetA(v float64) error
    GetB() float64
    SetB(v float64) error
    proto.Message
}

// Service interface with context support
type CalculatorServiceService interface {
    Add(ctx context.Context, req IAddRequest) (IAddResponse, error)
    Multiply(ctx context.Context, req IMultiplyRequest) (IMultiplyResponse, error)
    Divide(ctx context.Context, req IDivideRequest) (IDivideResponse, error)
}
```

## Test Plan

- [x] Unit tests for generator implementation
- [x] Proto file examples compile and generate expected interfaces
- [x] Generated code passes Go compilation
- [x] buf lint passes on example proto files
- [ ] Manual testing of generated interfaces with mock implementations
- [ ] Integration test with actual service implementation

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -2,6 +2,15 @@ module github.com/protomcp/protomcp
 
 go 1.23.0
```
[the remark this answers is not in the document]
Comment (amery) [reply go.mod:4]: your knowledge is outdated

**Comment by amery** [comment]
## Changes in this update

This update includes important fixes and improvements based on code review:

### 🐛 Bug Fix: Message Field Types
- **Fixed**: Message fields now correctly use interface types instead of pointer-to-interface
  - Before: `GetUser() *IUserInfo` ❌
  - After: `GetUser() IUserInfo` ✅
- **Fixed**: NoImpl stubs now correctly return `nil` for interface-typed fields (not zero struct)
- **Added**: Comprehensive test coverage for message field type handling

### 🧪 Test Improvements
- Added `makeContent()` helper methods to test cases for better code reuse
- Added specific test suite for message field types (`generator_message_fields_test.go`)
- Enhanced test utilities with:
  - `GetGeneratedFileContent` helper
  - `NewPlugin` factory function
  - Enum field support

### 📚 Documentation Updates
- Added field alignment instructions to AGENT.md for struct optimization
- Included example of using `fieldalignment` tool for better memory efficiency

### 🔧 Template Enhancements
- Updated NoImpl message template to detect interface fields and return `nil` appropriately
- Added `IsMessage` field to `FieldData` struct for template logic

### 🎯 Code Quality
- All tests pass with proper interface handling
- Generated code follows Go best practices for interfaces
- Consistent naming patterns across all generated types

The generated code now correctly handles all protobuf field types with proper Go semantics!

File: cmd/protoc-gen-protomcp/generator_template.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	var buf bytes.Buffer
+	if err := templates.ExecuteTemplate(&buf, fileTemplate, data); err != nil {
+		return fmt.Errorf("failed to execute template: %w", err)
+	}
+
+	// Write to generated file
+	genFile.P(buf.String())
+
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (amery) [reply cmd/protoc-gen-protomcp/generator_template.go]: is this truly obsolete?

**Comment by amery** [comment]
## 🔄 Major Update: Proto-Independent Interface Design

This update significantly improves the generator architecture and brings major design improvements that make the generated code more Go-idiomatic and production-ready.

## 🎯 Key Improvements

### 1. Proto-Independent Interface Design
- **BREAKING**: Removed `proto.Message` from generated interfaces
- **Benefit**: Generated interfaces are now pure Go with no protobuf dependencies
- **Impact**: Cleaner API surface, easier testing, reduced coupling

**Before:**
```go
type IAddRequest interface {
    GetA() float64
    SetA(v float64) error
    proto.Message  // ❌ Protobuf dependency
}
```

**After:**
```go  
type IAddRequest interface {
    GetA() float64
    SetA(v float64) error
    // ✅ Pure Go interface
}
```

### 2. Enhanced Build System
- **NEW**: `examples-build` target with formatting validation
- **IMPROVEMENT**: `examples` target now fails on unformatted code
- **CI-Ready**: Build fails if generated code isn't properly formatted

```bash
# New workflow
make examples        # Generate + validate formatting
make examples-build  # Generate + validate + build (CI-ready)
```

### 3. Comprehensive Testing Enhancements
- **Added**: 600+ lines of enhanced assertion utilities
- **Improved**: Better test organization with named test case types  
- **Enhanced**: More robust proto descriptor factories
- **Coverage**: Better edge case handling and error scenarios

### 4. Architecture Refinements
- **Options System**: Better nil-safe handling with comprehensive defaults
- **Import Optimization**: Smarter detection of required imports
- **Template Cleanup**: Removed unnecessary protobuf imports
- **Code Quality**: Cleaner generated code with optimized imports

## 📊 Quantified Changes

- **+904 lines, -200 lines** of improvements
- **21 files** updated across generator, tests, and examples
- **Enhanced test utilities** with comprehensive assertion helpers
- **Cleaner templates** removing protobuf dependencies
- **Improved build system** with validation steps

## 🔧 Migration Impact

### For Generated Code Users:
- **No Action Required**: Interfaces work the same, just cleaner
- **Testing Easier**: No need to mock protobuf dependencies
- **Imports Cleaner**: Fewer dependencies in generated files

### For Build Systems:
- **Validation Added**: Format checking now enforced in examples
- **CI Integration**: `examples-build` perfect for CI pipelines
- **Quality Gates**: Build fails on formatting issues

## 🚀 Production Readiness Improvements

1. **Dependency Reduction**: Generated interfaces are protobuf-free
2. **Build Validation**: Comprehensive format and build checking
3. **Test Coverage**: Enhanced assertion utilities for robust testing
4. **Code Quality**: Optimized imports and cleaner template output
5. **Developer Experience**: Better error messages and validation

## 🔍 Technical Details

### Template System Updates
- Removed `proto.Message` from message interfaces
- Removed `ProtoMessage()` from NoImpl stubs  
- Optimized import generation logic
- Cleaner template structure

### Testing Framework Evolution
- Enhanced assertion utilities with format string support
- Better proto descriptor factory functions
- Improved test organization patterns
- More comprehensive edge case coverage

### Build System Enhancement
- Format validation integrated into examples generation
- New `examples-build` target for CI workflows
- Better error reporting for format violations
- Consistent development experience

This update represents a significant maturation of the codebase, moving from prototype to production-ready generator with clean architecture and comprehensive validation.

File: README.md [hunk 3]
```diff
@@ -20,15 +20,81 @@ defined in Protocol Buffer (`.proto`) files, enabling them to serve both
...
+    GetB() float64
+    SetB(v float64) error
+    proto.Message
+}
+
+type ICalculatorService interface {
+    Add(ctx context.Context, req IAddRequest) (IAddResponse, error)
+}
+```
+
```
[the remark this answers is not in the document]
Comment (amery) [reply README.md:75]: great catch!

File: cmd/protoc-gen-protomcp/generator_template.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
```
[the remark this answers is not in the document]
Comment (amery) [reply cmd/protoc-gen-protomcp/generator_template.go:52]: opts can be nil, the accessors check for it

**Comment by amery** [comment]
## 🔄 Branch Update - Minor Refinements

Updated the local branch with **4 small improvements** that enhance code quality and documentation:

### 📝 Documentation Cleanup

- **README.md**: Removed `proto.Message` embedding from interface examples
  - Shows cleaner, pure Go interfaces without protobuf dependencies
  - Better demonstrates the proto-independence design goal

### 🧪 Test Code Organization

- **generator_test.go**: Extracted `generateFiles` helper function
  - Improved test code reusability and organization
  - No functional changes, just cleaner structure

### 🔤 Development Tools

- **cspell.json**: Added "Wrapf" to spell check dictionary
  - Supports the new `core.Wrapf` error handling pattern used in recent commits

### 🛠️ Code Simplification

- **helpers.go**: Minor refactoring in type generation logic
  - Simplified conditional branching for field type handling
  - Better code comments and readability
  - Removed redundant optional field handling

**Impact**: These are purely **quality-of-life improvements** that don't change core functionality. All features and APIs remain identical, just with cleaner code and better documentation examples.

The branch is now ready with all recent enhancements while maintaining full backward compatibility.
